shell/common/rasterizer.cc

```diff
@@ -172,8 +172,24 @@ void Rasterizer::Draw(
         if (discardCallback(*layer_tree.get())) {
           raster_status = RasterStatus::kDiscarded;
         } else {
-          raster_status =
-              DoDraw(std::move(frame_timings_recorder), std::move(layer_tree));
+          if (!layer_tree || !surface_) {
```

Comment:
Leave this to DoDraw, this creates an early return that didn't exist previously.

Reply:
The early return here is used to ensure that surface_ is not null pointer, because we need to use it right away 
```
if (surface_->AllowsDrawingWhenGpuDisabled()) {
  raster_status = DoDraw(std::move(frame_timings_recorder),
                         std::move(layer_tree));
```

Reply:
Good call, lets change it to:
```c++
if (!layer_tree || !surface_) {
  raster_status = RasterStatus::kFailed;
} else if (surface_->AllowsDrawingWhenGpuDisabled()) {
  raster_status = DoDraw(...);
} else {
  ...
}
```

This removes the new early out.  You can remove the check inside of DoDraw too now that it is here (replace it with asserts is probably better).

Reply:
Done. I’m not sure if assertions can be used here, so I’ll just keep the original logic unchanged.

Reply:
(although this becomes moot if we make the change I just suggested in the other comment, moving the sync switch to DrawToSurface)